### PR TITLE
[3679] - Subject change notification

### PR DESCRIPTION
**Pull request by TobyRet**
### Context
When a user updates a courses subjects we want to trigger a new 'Subjects updated' notification

### Changes proposed in this pull request
- 'Subjects updated' notification is triggered when the V2 API receives a request to update a course's subjects
- Any number of subjects can be updated

### 

### Guidance to review
- To test locally you will need to be subscribed to notifications and have a Notify key
- Go to a course that you are subscribed to get course updates for. Edit the courses subjects via 'basic details` tab. Once the update is submitted a subjects updated notification should fire.
- See a real example of the notification on Notify - https://www.notifications.service.gov.uk/services/022acc23-c40a-4077-bbd6-fc98b2155534/notification/1b579e0d-fd6b-4cf9-9806-321cd716c8f1

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

File: app/controllers/api/v2/courses_controller.rb [hunk 1]
```diff
@@ -102,6 +102,12 @@ def update
           @course.save
           @course.course_subjects.each(&:save)
 
+          NotificationService::CourseSubjectsUpdated.call(
+            course: @course,
+            previous_subject_names: @previous_subject_names,
```
Comment (TobyRet) [inline app/controllers/api/v2/courses_controller.rb:107]: Yep. Instance vars. WTF you may ask?
In the notification we need to show the user what the course subjects have changed from and to. Same deal with the course title. When a course's subjects change the course title changes dynamically to reflect the new subjects. We cannot use Rails Dirty API to (easily) to check if course associations have changed and how.

File: app/controllers/api/v2/courses_controller.rb [hunk 2]
```diff
@@ -176,8 +182,10 @@ def update_subjects
         if request_has_duplicate_subject_ids?
           @course.errors.add(:subjects, :duplicate)
         else
-          @course.subjects = []
+          @previous_subject_names = subject_names_for(@course)
```
Comment (TobyRet) [inline app/controllers/api/v2/courses_controller.rb:185]: Also worth noting that the `update_subjects` method is being called via the `create` action causing further complications / nightmares

File: app/mailers/course_subjects_updated_email_mailer.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+private
+
+  def format(subject_names)
+    if subject_names.length == 1
+      subject_names.first
+    else
+      subject_names.join(", ")
+    end
```
Comment (asmega) [inline app/mailers/course_subjects_updated_email_mailer.rb:34]: suggestion```
      subject_names.join(", ")
```
Does this not work? Why do you need to treat a collection of 1 different to a collection greater than 1?

File: app/services/notification_service/course_subjects_updated.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      @course = course
+      @previous_subject_names = previous_subject_names
+      @previous_course_name = previous_course_name
+    end
+
+    def call
+      return false unless notify_accredited_body?
+      return false unless course.in_current_cycle?
+
+      users = User.joins(:user_notifications).merge(UserNotification.course_update_notification_requests(course.accredited_body_code))
```
Comment (asmega) [inline app/services/notification_service/course_subjects_updated.rb:19]: you can extract this out to a private method instead of using a local variable

File: app/mailers/course_subjects_updated_email_mailer.rb (hunk 3 above)
Comment (TobyRet) [reply app/mailers/course_subjects_updated_email_mailer.rb:34]: Over-site my my part!

File: app/services/notification_service/course_subjects_updated.rb (hunk 4 above)
Comment (TobyRet) [reply app/services/notification_service/course_subjects_updated.rb:19]: Could do but what in way would doing so improve things? I would _extract method_ if:

- doing so made the code more readable (we already have a users variable and it already seems clear that we are pulling users from the db?)
- There was more than one place in the class that was using the same logic. In this case there isn't.

**Comment by defong** [comment]
[[S] Basic detail notification: Subject change](https://trello.com/c/RXZ1AVEO/3679-s-basic-detail-notification-subject-change)

File: app/controllers/api/v2/courses_controller.rb [hunk 5]
```diff
@@ -102,6 +102,14 @@ def update
           @course.save
           @course.course_subjects.each(&:save)
 
+          if subject_ids
+            NotificationService::CourseSubjectsUpdated.call(
+              course: @course,
+              previous_subject_names: @previous_subject_names,
+              previous_course_name: @previous_course_name,
```
Comment (aliuk2012) [inline app/controllers/api/v2/courses_controller.rb:109]: Not strictly related to this line of code but why do you need to use the previous course name?

Comment (aliuk2012) [reply app/controllers/api/v2/courses_controller.rb:109]: Might be worth checking how @avinhurry & I did course title changed notification https://github.com/DFE-Digital/teacher-training-api/pull/1481. Could this approach perhaps be a better on instead of having another notification template, azure changes etc.

Comment (TobyRet) [reply app/controllers/api/v2/courses_controller.rb:109]: As discussed - in the controller the courses subjects are being re-assigned (determined by what ever subject ids are passed in params). See https://github.com/DFE-Digital/teacher-training-api/pull/1478/files#diff-7d6b888ba704f4684c5b576a1dd07676R190

When updating subjects on a course the name is automatically updated to reflect the change in subjects. This is why the previous course name is being captured in an instance var and passed through to the service and mailer.

File: app/mailers/course_subjects_updated_email_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    set_template(Settings.govuk_notify.course_subjects_updated_email_template_id)
+
+    set_personalisation(
+      provider_name: course.provider.provider_name,
+      course_code: course.course_code,
+      subject_change_datetime: gov_uk_format(course.updated_at),
+      course_url: create_course_url(course),
+      previous_subjects: format(previous_subject_names),
+      updated_subjects: format(course.subjects.map(&:subject_name)),
```
Comment (misaka) [inline app/mailers/course_subjects_updated_email_mailer.rb:19]: Makes you wish there was a `course.subjects.subject_names` or something. :)

File: config/settings/test.yml [hunk 7]
```diff
@@ -5,6 +5,7 @@ govuk_notify:
   course_publish_email_template_id: please_change_me
   course_withdraw_email_template_id: please_change_me
   course_vacancies_filled_email_template_id: please_change_me
+  course_subjects_updated_email_template_id: please_change_me
```
Comment (misaka) [inline config/settings/test.yml:8]: @aliuk2012 recently demonstrated that there's no need to keep these template IDs secret. Can we start to just put them into the config here? And we'll have a larger ticket later to change the remaining ones?

Comment (aliuk2012) [reply config/settings/test.yml:8]: They should be in the global `settings.yml` file so they apply across all environments

File: azure/template.json [hunk 8]
```diff
@@ -199,6 +199,13 @@
         "description": "The GOV.UK Notify template id for the magic signin link"
       }
     },
+    "govukNotifyCourseSubjectsUpdatedEmailTemplateId": {
+      "type": "string",
+      "defaultValue": "",
+      "metadata": {
+        "description": "The GOV.UK Notify template id for the course subjects updated email"
+      }
+    },
```
Comment (misaka) [inline azure/template.json:208]: We should be able to get rid of these once we add these to the settings.

File: config/settings.yml [hunk 9]
```diff
@@ -13,6 +13,7 @@ govuk_notify:
   magic_link_email_template_id: please_change_me
   course_withdraw_email_template_id: please_change_me
   course_vacancies_filled_email_template_id: please_change_me
+  course_subjects_updated_email_template_id: please_change_me
```
Comment (misaka) [inline config/settings.yml:16]: The comment on `config/settings/test.yml` should be here, can we put the real templated IDs in here pls.

File: config/settings/test.yml (hunk 7 above)
Comment (misaka) [reply config/settings/test.yml:8]: Do we need these in the test env still?

File: app/services/notification_service/course_subjects_updated.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        ).deliver_later(queue: "mailer")
+      end
+    end
+
+  private
+
+    attr_reader :course, :previous_subject_names, :previous_course_name
+
+    def users
+      @users ||= User.joins(:user_notifications).merge(UserNotification.course_update_notification_requests(course.accredited_body_code))
```
Comment (misaka) [inline app/services/notification_service/course_subjects_updated.rb:34]: Would it make sense to move this into `User`? Normally it makes sense for a service to coordinate between multiple models, but in this case the `UserNotification` is kind of an extension to `User`.

Comment (TobyRet) [reply app/services/notification_service/course_subjects_updated.rb]: Good idea @misaka 
Done -> https://github.com/DFE-Digital/teacher-training-api/pull/1478/commits/3b48730bd9a2f20d10ae0f1d5374ac8016d1a2ed

File: spec/services/notification_service/course_subjects_updated_spec.rb [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        it "does not send a notification" do
+          expect(CourseSubjectsUpdatedEmailMailer).to_not receive(:course_subjects_updated_email)
+          expect(course.recruitment_cycle).to_not eql(RecruitmentCycle.current)
+          described_class.call(
+            course: course,
+            previous_subject_names: previous_subject_names,
+            previous_course_name: previous_course_name,
+          )
+        end
+      end
```
Comment (misaka) [inline spec/services/notification_service/course_subjects_updated_spec.rb:85]: I'm finding these tests really hard to read. Now that I've been staring at them for a bit I feel like I can kind of understand what's being tested, but I can't help but wonder if it wouldn't be better organised if we used the pattern where we do call the service in a before block, and then do the setup and testing separately. Something like:

```ruby
    describe "#call" do
      # all the let blocks

      before do
        setup_notifications

        described_class.call(
          course: course,
          previous_subject_names: subject_names,
          previous_course_name: course_name,
        )
      end

      describe CourseSubjectsUpdatedEmailMailer do
        before do
          allow(CourseSubjectsUpdatedEmailMailer).to receive(:course_subjects_updated_email)
        end

        context "with a course that is in the current cycle" do
          it { is_expected.to receive(:course_subjects_updated_email) }
        end

        context "with a course that is not in the current cycle" do
          let(:provider) { create(:provider, :next_recruitment_cycle) }
          let(:course) { create(:course, accredited_body_code: accredited_body.provider_code, provider: provider) } 

          it { is_expected.to_not receive(:course_subjects_updated_email) }
        end
      end

      # and the rest of the specs modified similarly
```

Comment (misaka) [reply spec/services/notification_service/course_subjects_updated_spec.rb:85]: Alternatively, it seems that this code is used each time to trigger the work needed for the test:

```ruby
described_class.call(
  course: course,
  previous_subject_names: previous_subject_names,
  previous_course_name: previous_course_name,
)
```

I find it's hard to parse that while also keeping all the context in my head ... so maybe turn that into a method that makes is clear that it's the same call throughout:

```ruby
def call_service
  described_class.call(
    course: course,
    previous_subject_names: previous_subject_names,
    previous_course_name: previous_course_name,
  )
end


it "does not send a notification" do
  expect(CourseSubjectsUpdatedEmailMailer).to_not receive(:course_subjects_updated_email)
  expect(course.recruitment_cycle).to_not eql(RecruitmentCycle.current)
  call_service
end 
```

Comment (TobyRet) [reply spec/services/notification_service/course_subjects_updated_spec.rb:85]: Yep cool - have made the changes
